### PR TITLE
Jetpack Onboarding: Setup initial structure

### PR DESCRIPTION
**Pull request by tyxla**
This PR provides the basic structure for the Jetpack Onboarding project:

* Provide a feature flag, enabled only on the `development` environment
* Allow `Wizard` steps to retrieve back/forward URL (addressed separately in #20402)
* Add a main wrapper component for the steps logic
* Define the steps and their components
* Create the Jetpack Onboarding section and its routes and controller.
* Add all steps as empty components with just title and subtitle.

Based on the following initial prototype PR: #20360

To test:
* Checkout this branch
* Go to http://calypso.localhost:3000/jetpack/onboarding/
* Go through back and forth through all steps and verify you can navigate properly.
* Try visiting each of the steps in a new window and verify you can access them.

### CODE REVIEW

**Comment by AnnaMag** [comment]
Tested 👍 Thank you!

I haven't looked at the code in much detail, but it seems like a good starting structure to work off of. Since it's behind the feature flag, should we merge it and adapt later if necessary?

**Comment by tyxla** [comment]
> Since it's behind the feature flag, should we merge it and adapt later if necessary?

That's my intention, let's also see if @ockham has any objections before we 🚢

File: client/jetpack-onboarding/constants.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+};
+
+export const JETPACK_ONBOARDING_COMPONENTS = {
+	[ JETPACK_ONBOARDING_STEPS.SITE_TITLE ]: <JetpackOnboardingSiteTitleStep />,
+	[ JETPACK_ONBOARDING_STEPS.SITE_TYPE ]: <JetpackOnboardingSiteTypeStep />,
+	[ JETPACK_ONBOARDING_STEPS.HOMEPAGE ]: <JetpackOnboardingHomepageStep />,
+	[ JETPACK_ONBOARDING_STEPS.CONTACT_FORM ]: <JetpackOnboardingContactFormStep />,
+	[ JETPACK_ONBOARDING_STEPS.WOOCOMMERCE ]: <JetpackOnboardingWoocommerceStep />,
+	[ JETPACK_ONBOARDING_STEPS.SUMMARY ]: <JetpackOnboardingSummaryStep />,
+};
```
Comment (ockham) [inline client/jetpack-onboarding/constants.js]: Do we need all this indirection? I'd just turn `JETPACK_ONBOARDING_COMPONENTS`' keys into string literals. If we ever need `JETPACK_ONBOARDING_STEPS` separately, we can define 'em as `Object.keys( JETPACK_ONBOARDING_COMPONENTS )`.

File: client/jetpack-onboarding/controller.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * Internal Dependencies
+ */
+import JetpackOnboardingMain from './main';
+import { renderWithReduxStore } from 'lib/react-helpers';
+import { setSection } from 'state/ui/actions';
+
+const removeSidebar = context => {
+	ReactDom.unmountComponentAtNode( document.getElementById( 'secondary' ) );
+	context.store.dispatch( setSection( null, { hasSidebar: false } ) );
+};
```
Comment (ockham) [inline client/jetpack-onboarding/controller.js]: Looking forward to when we'll be able to ditch this kind of multi-tree rendering helpers, i.e. when STR lands.

File: client/jetpack-onboarding/controller.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+export const onboarding = context => {
+	removeSidebar( context );
+
+	renderWithReduxStore(
+		React.createElement( JetpackOnboardingMain, {
+			stepName: context.params.stepName,
+		} ),
```
Comment (ockham) [inline client/jetpack-onboarding/controller.js]: Let's use JSX here -- it's sort of Calypso convention by now to prefer that over `createElement`.

File: client/jetpack-onboarding/index.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+import { isEnabled } from 'config';
+
+/**
+ * Internal dependencies
+ */
+import { onboarding } from './controller';
+
+export default function() {
+	if ( isEnabled( 'jetpack/onboarding' ) ) {
+		page( '/jetpack/onboarding/:stepName?', onboarding );
```
Comment (ockham) [inline client/jetpack-onboarding/index.js]: We could (and should 😄) add a whitelist for `stepName` here, cf #20422. Fine to do it in separate PR tho.

File: client/jetpack-onboarding/main.jsx [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+/**
+ * Internal dependencies
+ */
+import Wizard from 'components/wizard';
+import {
+	JETPACK_ONBOARDING_COMPONENTS as COMPONENTS,
+	JETPACK_ONBOARDING_STEPS as STEPS,
+} from './constants';
+
+class JetpackOnboardingMain extends React.Component {
```
Comment (ockham) [inline client/jetpack-onboarding/main.jsx]: Let's make this a `PureComponent`. (I know `connect()` will turn it into one anyway, but I like to explicitly state it whenever possible).

File: client/jetpack-onboarding/main.jsx [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+				steps={ steps }
+				stepName={ stepName }
+			/>
+		);
+	}
+}
+
+export default connect( () => {
+	// Note: here we can select which steps to display, based on user's input
+	const steps = compact( [
```
Comment (ockham) [inline client/jetpack-onboarding/main.jsx]: Why the `compact`? We're in charge of defining these constants, so there shouldn't be any falsy ones anway, no?

File: client/jetpack-onboarding/steps/business-address.jsx [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+import React from 'react';
+import { localize } from 'i18n-calypso';
+
+/**
+ * Internal dependencies
+ */
+import FormattedHeader from 'components/formatted-header';
+
+class JetpackOnboardingBusinessAddressStep extends React.Component {
+	render() {
```
Comment (ockham) [inline client/jetpack-onboarding/steps/business-address.jsx]: Let's make it either stateless/functional our a `PureComponent`.

In principle, I like to start with stateless whenever possible. We're arguable going to need state or event handlers in most of these steps, but usually, it's easier to forget to turn something that doesn't back into a stateless one.

`PureComponent` could be a good compromise.

File: client/jetpack-onboarding/constants.js (hunk 1 above)
Comment (tyxla) [reply client/jetpack-onboarding/constants.js]: I'm more inclined to leave them like that. We'll be referring to those steps and components in other places in the code, and it will generally be much safer to refer to them by using the constants.

File: client/jetpack-onboarding/controller.js (hunk 3 above)
Comment (tyxla) [reply client/jetpack-onboarding/controller.js]: Done in 54dc9a46291bc4e805428106545715ff7bb960f5

File: client/jetpack-onboarding/index.js (hunk 4 above)
Comment (tyxla) [reply client/jetpack-onboarding/index.js]: I'd be happy to defer that to you 😆 TBH I'm not sure what's the rationale behind that.

File: client/jetpack-onboarding/steps/business-address.jsx (hunk 7 above)
Comment (tyxla) [reply client/jetpack-onboarding/steps/business-address.jsx]: Done 👍 8547b85e6cade12ff5ea0a70b10c94b40bcfce8a

File: client/jetpack-onboarding/main.jsx (hunk 5 above)
Comment (tyxla) [reply client/jetpack-onboarding/main.jsx]: Done 👍 8547b85e6cade12ff5ea0a70b10c94b40bcfce8a

File: client/jetpack-onboarding/main.jsx (hunk 6 above)
Comment (tyxla) [reply client/jetpack-onboarding/main.jsx]: Ah, good question. 

The thing is that some steps might be enabled, others might not be enabled. This, in some cases, will depend on the user input (which we're planning to store in Redux). That's the rationale behind putting those in the `mapStateToProps`, essentially.

For example, the business address step will be shown only if the user has selected the Business site type. So for example, here we'll have:

```
getJetpackOnboardingSiteType( state ) === 'business' && STEPS.BUSINESS_ADDRESS
```

Let me know if that still doesn't make any sense.

Comment (ockham) [reply client/jetpack-onboarding/main.jsx]: I figured that was the reason, but I'd go with YAGNI for now. Easy enough to add that `compact` later.

File: client/jetpack-onboarding/index.js (hunk 4 above)
Comment (ockham) [reply client/jetpack-onboarding/index.js]: Okay, will do in a subsequent PR.

As for the rationale, we'll want to avoid matching routes for which we don't know what to do. Using a whitelist here, we can fall through to a 404 (or, if it's only on the client side, a simple 'Not Found' component) in all other cases.
